packages/@internationalized/date/src/manipulation.ts

```diff
@@ -115,7 +115,7 @@ function balanceDay(date: Mutable<AnyCalendarDate>) {
 
 function constrainMonthDay(date: Mutable<AnyCalendarDate>) {
   date.month = Math.max(1, Math.min(date.calendar.getMonthsInYear(date), date.month));
-  date.day = Math.max(1, Math.min(date.calendar.getDaysInMonth(date), date.day));
+  date.day = Math.max(1, Math.min(31, date.day));
```

Comment:
I don't think the util should be loose on the constraint given the name of the function, instead, wherever it is we need it to be loose, we should do this. This way it's obvious when we are truly constraining to a valid date vs dealing with partial information.